Program.cs

```diff
@@ -471,65 +480,127 @@ static P2pEndPoint GetExternalEndPoint(Socket socket)
+                Console.WriteLine("[{0}] - Waiting {1} sec to sync with other peer", now.ToLongTimeString(), sleepTimeToSync);
 
-                    if (client != null)
-                        client.Close();
+                await Task.Delay(TimeSpan.FromSeconds(sleepTimeToSync), cts.Token);
 
-                    client = new UdtSocket(socket.AddressFamily, socket.SocketType);
-                    client.Bind(socket);
+                Console.WriteLine("Connecting to other peer, please wait... ");
+                udtSocket = await (await Task.WhenAny(taskList));
+            } while (udtSocket == null);
 
-                    Console.Write("\r{0} - Trying to connect to {1}:{2}.  ",
-                        retry++, remoteAddr, remotePort);
+            cts.Cancel();
 
-                    client.Connect(new IPEndPoint(IPAddress.Parse(remoteAddr), remotePort));
+            return udtSocket;
+        }
+
+        static async Task<UdtSocket> ConnectAsync(Socket socket, string remoteAddress, int remotePort, CancellationToken ct)
+        {
+            try
+            {
+                await GetExternalEndPointAsync(socket);
+
+                var client = new UdtSocket(socket.AddressFamily, socket.SocketType);
+                client.Bind(socket);
+
+                Console.WriteLine("Trying to connect to {0}:{1}.  ", remoteAddress, remotePort);
+
+                client.Connect(new IPEndPoint(IPAddress.Parse(remoteAddress), remotePort));
 
-                    Console.WriteLine("Connected successfully to {0}:{1}",
-                        remoteAddr, remotePort);
+                Console.WriteLine("Connected successfully to {0}:{1}", remoteAddress, remotePort);
 
-                    bConnected = true;
+                if (ct.IsCancellationRequested)
+                {
+                    socket.Close();
+                    ct.ThrowIfCancellationRequested();
                 }
-                catch (Exception e)
+
+                return client;
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine($"{e.Message.Replace(Environment.NewLine, ". ")} - Port:{remotePort}");
+                return null;
+            }
+        }
+
+        static async Task InitializeSocketsAsync()
+        {
+            var socket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
+            socket.Bind(new IPEndPoint(IPAddress.Any, 0));
+
+            // Connecting to external to obtain IP and get firewall type
+            var p2pEndPoint = await GetExternalEndPointAsync(socket);
+            if (p2pEndPoint == null)
+            {
+                socket.Close();
+                return;
+            }
+
+            Console.WriteLine();
+            Console.WriteLine("Tell your peer your IP is: {0}", p2pEndPoint.External.Address);
+            Console.WriteLine("Initializing sockets...");
+
+            var ports = new StringBuilder();
+
+            for (var i = 0; i < 10; i++)

```

Comment:
can you store the number 10 in a constant so that it's easily findable and changeable later please

Reply:
Ok sure